File: sdks/node/src/resources/events.ts
import { ApiClient } from '../client';
import {
  DeleteEventParams,
  DeleteEventResponse,
  EventListParams,
  EventResponse,
  EventsListResponse,
  UpdateEventData,
  CreateEventData,
  ProviderName,
} from '../types';

/**
 * Events resource for managing calendar events across multiple providers
 */
export class Events {
  constructor(private readonly client: ApiClient) {}

  /**
   * List calendar events
   *
   * @param params - List parameters (start, end, calendarIds, etc.)
   */
  async list(params?: EventListParams): Promise<EventsListResponse> {
    const queryParams = new URLSearchParams();

    if (params) {
      const { pageSize, start, end, calendarIds, nextPageToken, appleToken, singleEvents } = params;

      if (pageSize) queryParams.append('pageSize', pageSize.toString());
      if (start) queryParams.append('start', start instanceof Date ? start.toISOString() : start);
      if (end) queryParams.append('end', end instanceof Date ? end.toISOString() : end);
      if (calendarIds) queryParams.append('calendarIds', JSON.stringify(calendarIds));
      if (nextPageToken) queryParams.append('nextPageToken', nextPageToken);
      if (appleToken) queryParams.append('appleToken', JSON.stringify(appleToken));
      if (singleEvents !== undefined) queryParams.append('singleEvents', singleEvents.toString());
    }

    const query = queryParams.toString();
    const path = query ? `/events?${query}` : '/events';

    const response = await this.client.get<EventsListResponse>(path);
    return response.data || { events: [] };
  }

  /**
   * Create a new calendar event
   *
   * @param params - Event creation parameters including provider
   */
  async create(params: CreateEventData & { provider: ProviderName }): Promise<EventResponse> {
    const response = await this.client.post<EventResponse>('/event', params);
    return response.data;
  }

  /**
   * Update an existing calendar event
   *
   * @param params - Event update parameters including provider
   */
  async update(params: UpdateEventData & { provider: ProviderName }): Promise<EventResponse> {
    const response = await this.client.put<EventResponse>('/event', params);
    return response.data;
  }

  /**
   * Delete a calendar event
   *
   * @param params - Event deletion parameters including provider
   */
  async delete(params: DeleteEventParams): Promise<DeleteEventResponse> {
    const queryParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined) {
        queryParams.append(key, value.toString());
      }
    });

    const query = queryParams.toString();
    const path = query ? `/event?${query}` : '/event';

    const response = await this.client.delete<DeleteEventResponse>(path);
    return response.data;
  }
}

File: sdks/dotnet/src/Mobiscroll.Connect/Resources/Events.cs
using System;
using System.Globalization;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Mobiscroll.Connect.Internal;
using Mobiscroll.Connect.Models;

namespace Mobiscroll.Connect.Resources;

public sealed class Events
{
    private readonly ApiClient _client;

    internal Events(ApiClient client)
    {
        _client = client;
    }

    /// <summary>List events matching the given filters.</summary>
    public async Task<EventsListResponse> ListAsync(EventListParams? listParams = null, CancellationToken ct = default)
    {
        var path = "/events";
        if (listParams is not null)
        {
            var qs = new QueryStringBuilder()
                .Add("pageSize", listParams.PageSize)
                .Add("start", FormatDate(listParams.Start))
                .Add("end", FormatDate(listParams.End))
                .Add("nextPageToken", listParams.NextPageToken)
                .Add("singleEvents", listParams.SingleEvents);

            if (listParams.CalendarIds is { Count: > 0 })
            {
                qs.Add("calendarIds", JsonSerializer.Serialize(listParams.CalendarIds, ApiClient.JsonOptions));
            }

            if (!qs.IsEmpty)
            {
                path = $"/events?{qs}";
            }
        }

        var result = await _client.GetAsync<EventsListResponse>(path, ct).ConfigureAwait(false);
        return result ?? new EventsListResponse();
    }

    /// <summary>Create a new event on the given provider+calendar.</summary>
    public async Task<CalendarEvent> CreateAsync(EventCreateData data, CancellationToken ct = default)
    {
        if (data is null)
        {
            throw new ArgumentNullException(nameof(data));
        }
        var result = await _client.PostJsonAsync<CalendarEvent>("/event", data, ct).ConfigureAwait(false);
        return result ?? throw new InvalidOperationException("Event creation returned no body");
    }

    /// <summary>Update an existing event.</summary>
    public async Task<CalendarEvent> UpdateAsync(EventUpdateData data, CancellationToken ct = default)
    {
        if (data is null)
        {
            throw new ArgumentNullException(nameof(data));
        }
        var result = await _client.PutJsonAsync<CalendarEvent>("/event", data, ct).ConfigureAwait(false);
        return result ?? throw new InvalidOperationException("Event update returned no body");
    }

    /// <summary>Delete an event (optionally a single recurrence instance).</summary>
    public Task DeleteAsync(EventDeleteParams deleteParams, CancellationToken ct = default)
    {
        if (deleteParams is null)
        {
            throw new ArgumentNullException(nameof(deleteParams));
        }

        var qs = new QueryStringBuilder()
            .Add("provider", deleteParams.Provider.ToWireString())
            .Add("calendarId", deleteParams.CalendarId)
            .Add("eventId", deleteParams.EventId)
            .Add("recurringEventId", deleteParams.RecurringEventId)
            .Add("deleteMode", deleteParams.DeleteMode);

        var path = qs.IsEmpty ? "/event" : $"/event?{qs}";
        return _client.DeleteAsync(path, ct);
    }

    private static string? FormatDate(DateTime? value)
    {
        if (!value.HasValue) return null;
        var dt = value.Value;
        var utc = dt.Kind == DateTimeKind.Utc ? dt : dt.ToUniversalTime();
        return utc.ToString("yyyy-MM-ddTHH:mm:ss.fffZ", CultureInfo.InvariantCulture);
    }
}
